packages/next/build/webpack/plugins/next-trace-entrypoints-plugin.ts

```diff
@@ -0,0 +1,238 @@
+
+      entrypoint.getFiles().forEach((file: string) => {
+        if (!file.endsWith(entrypoint.name + '.js')) {
+          entryFiles.add(nodePath.join(compilation.outputOptions.path, file))
+        }
+      })
+
+      for (const block of entryMod.blocks) {
+        const chunkGroup = getChunkGroupFromBlock(compilation, block)
+
+        if (!chunkGroup) {
+          continue
+        }
+
+        for (const chunk of (chunkGroup as any).chunks || []) {
+          chunk.files?.forEach((file: string) => {
+            entryFiles.add(nodePath.join(compilation.outputOptions.path, file))
+          })
+        }
+      }
+
+      assets[
+        `${isWebpack5 ? '../' : ''}${entrypoint.name}.nft.json`
+      ] = new sources.RawSource(
+        JSON.stringify({
+          version: TRACE_OUTPUT_VERSION,
+          files: [...entryFiles, ...this.entryTraces.get(entrypoint.name)!],
+        })
+      )
+    }
+  }
+
+  apply(compiler: webpack.Compiler) {
+    if (isWebpack5) {
+      compiler.hooks.make.tap('NextJsPagesManifest', (compilation) => {
+        // @ts-ignore TODO: Remove ignore when webpack 5 is stable
+        compilation.hooks.processAssets.tap(
+          {
+            name: 'NextJsPagesManifest',
+            // @ts-ignore TODO: Remove ignore when webpack 5 is stable
+            stage: webpack.Compilation.PROCESS_ASSETS_STAGE_ADDITIONS,
+          },
+          (assets: any) => {
+            this.createTraceAssets(compilation, assets)
+          }
+        )
+      })
+    } else {
+      compiler.hooks.emit.tap('NextJsPagesManifest', (compilation: any) => {
+        this.createTraceAssets(compilation, compilation.assets)
+      })
+    }
+
+    compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
+      compilation.hooks.finishModules.tapAsync(
+        PLUGIN_NAME,
+        async (_stats: any, callback: any) => {
+          try {
+            const depModMap = new Map<string, any>()
+
+            compilation.entries.forEach((entry) => {
+              const name = entry.name || entry.options?.name
+
+              if (name?.startsWith('pages/') && entry.dependencies[0]) {
+                const entryMod = getModuleFromDependency(
+                  compilation,
+                  entry.dependencies[0]
+                )
+
+                if (entryMod.resource) {
+                  this.entryNameMap.set(entryMod.resource, name)
+                  this.entryModMap.set(entryMod.resource, entryMod)
+                }
+              }
+            })
+
+            const readFile = (path: string) => {
+              const mod = depModMap.get(path) || this.entryModMap.get(path)
+
+              // map the transpiled source when available to avoid
+              // parse errors in node-file-trace
+              if (mod?._source?._valueAsBuffer) {
+                return mod._source._valueAsBuffer
+              }
+
+              try {
+                return compilation.inputFileSystem.readFileSync(path)
```

Comment:
Reading the files directly from fs is probably not the best idea.

A loader could have modified the source code after filesystem. Or it could be transpiled from a language nft doesn't understand.

There is a `NormalModule.originalSource()` function which gives you the source code for a module after loader processing. That's in a language webpack understands, so either javascript, or something else you can ignore. Best check `Module.type` for `javascript/*` to only process JS code.

Note that some modules might not end up at all in the output, so best use the list of chunks from `createTraceAssets` and grab all modules from there, de-duplicate them, analyse them with nft for more references, merge and cache the results per chunk and add them to the analysis data while iterating over the chunks.

Reply:
It looks like we can't use the list of files from `createTraceAssets` during this hook since we need it to be immediately after transpiling with babel and `compilation.entrypoints.values()` seems to be empty at this stage. I added some additional tests and gathering the files from the entry's module dependencies seems to gather everything we need currently and allows us to use the dependencies source if available as well.